### DISCUSSION TITLE
What is the best way to define other data types at emulator?

### DISCUSSION THREAD

**Discussion by Akay7** [pymodbus-dev/pymodbus#2429] [Q&A] [answered]
I'm trying to emulate device with Modbus TCP connection. And I found that there's no defined types for:
* int32
* uint64 
* int64

although [uint32, float32 already defined](https://github.com/pymodbus-dev/pymodbus/blob/69dc58ed9f811080737365850e30b52b7a64c244/examples/datastore_simulator_share.py#L79-L109).

What is the best way to define such types for device emulator? Is it better to start with single registers or there's some shortcuts or examples?

**Comment by janiversen** [comment]
The best way is to make a pull request that adds these types.

int32 are already covered by uinit32....there are no protocol difference between int and uint, so no need to have both types.

**Reply by Akay7** [reply]
> int and uint, so no need to have both types.

But what about large numbers? First bit is represent that is a negative number. For example in C++ 

type | possible values
--- | ----
int32 |  −2,147,483,648 to 2,147,483,647 
uint32 |	0 to 4,294,967,295

Maybe I missing something here. but I think that int32 doesn't covered by uint32 at this package.

**Reply by janiversen** [reply, marked as answer]
Well you might not think so, but you forget the the number if converted to registers each 16 bit, so it does not matter.

It is correct that when you define the values in the json file there is a difference, but that is really the only difference, and not to forget it complicated the configuration....do you have a real use case, where you need to define large integers ?

As I wrote earlier, if you want changes please submit a change request.

**Reply by Akay7** [reply]
> It is correct that when you define the values in the json file there is a difference, but that is really the only difference, and not to forget it complicated the configuration....do you have a real use case, where you need to define large integers ?

Yeah, there's a power meter with some information stored at uint64, some at int64. And it could be nice if I will able to emulate and check end-to-end edge cases.

> As I wrote earlier, if you want changes please submit a change request.

Thanks @janiversen ! I though that already exist since someone use that project and that my issue that I don't see that information in docs. But now I know that this functionality does not exists and PRs are welcomed.

**Reply by janiversen** [reply]
There are others using the simulator, when I specify 64 bit values, I use my calculator, convert it to hex, and add the hex value to each of the 4 registers.....with negative values I just ask the calculator for the unsigned value.

That said, it would be a nice enhancement to support uint64.
